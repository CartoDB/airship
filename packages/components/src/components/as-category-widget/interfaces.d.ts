export interface Category {
  name: string;
  value: number;
  color?: string;
}
export interface CategoryOptions {
  maximumValue: number;
  weight?: number | string;
  isOther?: boolean;
}
