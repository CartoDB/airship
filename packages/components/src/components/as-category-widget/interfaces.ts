export interface Category {
  name: string;
  value: number;
  color?: string;
}
export interface CategoryOptions {
  maximumValue: number;
  isOther?: boolean;
}
