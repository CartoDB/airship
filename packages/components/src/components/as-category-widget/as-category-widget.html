<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <base href="/">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
    <title>Airship</title>
    <link rel="stylesheet" href="/dist/airship.css">
    <script src="/build/airship.js"></script>
  </head>

  <body>
    <as-category-widget
      class="as-p--16"
      heading="Business Volume"
      description="Description"></as-category-widget>

    <script>
      var categoryWidget = document.querySelector('as-category-widget');
      categoryWidget.showClearButton = true;
      categoryWidget.categories = [
        { name: 'Bars & Restaurants', value: 1000, color: '#FABADA' },
        { name: 'Fashion', value: 900 },
        { name: 'Grocery', value: 800 },
        { name: 'Health', value: 400 },
        { name: 'Shopping mall', value: 250 },
        { name: 'Transportation', value: 650 }
      ];
      categoryWidget.addEventListener('categoriesSelected', function(event) {
        console.log('Categories Selected', event.detail)
      });
    </script>
  </body>
</html>
