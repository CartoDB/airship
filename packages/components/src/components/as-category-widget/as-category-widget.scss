@import '../../../../styles/src/core/_dev';

$bar--height: get-spacing(1);

as-category-widget {
  --category-widget--description--color: #{$color-type-02};
  --category-widget--bar--height: #{$bar--height};
  --category-widget--background-color: #{$color-ui-01};
  --category-bar--background-color: #{$color-ui-02};
  --category-bar--color: #{$color-complementary}; // overwritten by the default-bar-color attribute!

  display: block;
  background: var(--category-widget--background-color, $color-ui-01);

  .as-category-widget__heading {
    margin: 0;
    font: $font-subheader;
  }

  .as-category-widget__description,
  .as-category-widget__count {
    color: var(--category-widget--description--color, $color-type-02);
    color: $color-type-02;
  }

  .as-category-widget__list {
    margin: 0;
    padding: 0;
    list-style: none;

    &.as-category-widget__list--disabled {
      pointer-events: none;
    }
  }

  .as-category-widget__footer {
    display: flex;
    justify-content: space-between;
  }

  .as-category-widget__description + .as-category-widget__list {
    margin-top: get-spacing(4);
  }

  .as-category-widget__list + .as-category-widget__footer {
    margin-top: get-spacing(4);
  }

  .as-category-widget__info {
    display: flex;
    justify-content: space-between;
  }

  .as-category-widget__title {
    flex: 1;
    width: 100%;
    padding-right: get-spacing(2);
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .as-category-widget__bar {
    position: relative;
    width: 100%;
    height: $bar--height;
    border-radius: get-spacing(0.5);
    background-color: var(--category-bar--background-color, $color-ui-02);
    background-color: $color-ui-02;
  }

  .as-category-widget__bar-value {
    position: absolute;
    left: 0;
    max-width: 100%;
    height: $bar--height;
    transition: background 0.2s ease, opacity 0.5s ease;
    border-radius: get-spacing(0.5);
  }

  .as-category-widget__category {
    margin-bottom: get-spacing(2);
    cursor: pointer;

    &--other {
      pointer-events: none;
    }

    &--not-selected {
      opacity: 0.5;
    }

    &:not(.as-category-widget__category--selected):hover {
      .as-category-widget__bar-value {
        opacity: 0.6;
      }
    }
  }
}
