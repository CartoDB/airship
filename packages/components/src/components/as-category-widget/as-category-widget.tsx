import { Component, Element, Event, EventEmitter, Method, Prop, State } from '@stencil/core';
import readableNumber from '../../utils/readable-number';
import { shadeOrBlend } from '../../utils/styles';

const OTHER_CATEGORY_COLOR = '#747474';
const OTHER_CATEGORY_NAME = 'Other';

/**
 * Category Widget
 *
 * @export
 * @class CategoryWidget
 */
@Component({
  shadow: false,
  styleUrl: './as-category-widget.scss',
  tag: 'as-category-widget',
})
export class CategoryWidget {
  /**
   * Array of categories to display in the widget.
   * Each category should include a `name` and a `value`.
   * You can also override the bar color for each category with `color`.
   *
   * @type {object[]}
   * @memberof CategoryWidget
   */
  @Prop() public categories: object[] = [];

  /**
   * Default color to draw the bars. Default value is `#47DB99`.
   *
   * @type {string}
   * @memberof CategoryWidget
   */
  @Prop() public defaultBarColor: string;

  /**
   * Description text of the widget
   *
   * @type {string}
   * @memberof CategoryWidget
   */
  @Prop() public description: string;

  /**
   * Disable category selection in Widget
   *
   * @type {string}
   * @memberof CategoryWidget
   */
  @Prop() public disableInteractivity: boolean = false;

  /**
   * Heading text of the widget
   *
   * @type {string}
   * @memberof CategoryWidget
   */
  @Prop() public heading: string;

  /**
   * If truthy, it'll show a button to clear selected categories when there are any. Default value is `false`.
   *
   * @type {boolean}
   * @memberof CategoryWidget
   */
  @Prop() public showClearButton: boolean = false;

  /**
   * If truthy, it'll render the heading and component's description. Default value is `true`.
   *
   * @type {boolean}
   * @memberof CategoryWidget
   */
  @Prop() public showHeader: boolean = true;

  /**
   * If truthy, we'll use the sum of all categories' value to render the bar percentage.
   * By default, we use the maximum category value to render the bar percentage.
   *
   * @type {boolean}
   * @memberof CategoryWidget
   */
  @Prop() public useTotalPercentage: boolean = false;

  /**
   * The number of visible categories without aggregation.
   *
   * @type {number}
   * @memberof CategoryWidget
   */
  @Prop() public visibleCategories: number = Infinity;

  /**
   * Fired when selected categories changed or selected categories are cleared.
   *
   * @event categoriesSelected
   * @type {EventEmitter<string[]>}
   * @memberof CategoryWidget
   */
  @Event() public categoriesSelected: EventEmitter<string[]>;

  @State() private selectedCategories: string[] = [];

  @Element() private el: HTMLElement;

  /**
   * Get current selected categories
   *
   * @returns
   * @memberof CategoryWidget
   */
  @Method()
  public getSelectedCategories() {
    return this.selectedCategories;
  }

  public componentWillLoad() {
    this.el.style.setProperty('--category-bar-color', this.defaultBarColor || Â `var(--as-color-complementary)`);
  }

  /**
   * Clear current selected categories
   *
   * @returns
   * @memberof CategoryWidget
   */
  @Method()
  public clearSelection() {
    if (!this.selectedCategories.length) {
      return;
    }

    this.selectedCategories = [];
    this._onCategoriesChanged();
  }

  public render() {
    return [
      this._renderHeader(),
      this._renderCategoryList(),
      !this.disableInteractivity ? this._renderFooter() : ''
    ];
  }

  private _renderHeader() {
    if (!this.showHeader) {
      return;
    }

    return [
      <h2 class='as-category-widget__heading'>{this.heading}</h2>,
      <p class='as-category-widget__description as-body'>{this.description}</p>,
    ];
  }

  private _renderCategoryList() {
    const cssClasses = {
      'as-category-widget__list': true,
      'as-category-widget__list--disabled': this.disableInteractivity
    };

    return <ul class={cssClasses}>{this._renderCategories()}</ul>;
  }

  private _renderCategories() {
    const moreCategoriesThanVisible = this.categories.length > this.visibleCategories;
    const { categories, otherCategory } = this._parseCategories();
    let otherCategoryTemplate;

    const categoriesToRender = categories.slice(0, this.visibleCategories);

    const maximumValue = this.useTotalPercentage
      ? this._getCategoriesTotalValue(this.categories)
      : this._getCategoriesMaximumValue(categories, Boolean(otherCategory));

    if (otherCategory || moreCategoriesThanVisible) {
      otherCategoryTemplate = this._renderOtherCategory(otherCategory, { maximumValue });
    }

    return [
      categoriesToRender.map((category: Category) => this._renderCategory(category, { maximumValue })),
      otherCategoryTemplate
    ];
  }

  private _renderCategory(category: Category, options: CategoryOptions) {
    const { isOther, maximumValue } = options;
    const isSelected = this._isSelected(category.name);
    const isAnyCategorySelected = this.selectedCategories.length > 0;

    const barColor = this._getBarColor(category.color, { isSelected, isOther });

    const progressStyles = {
      backgroundColor: barColor ? barColor : `var(--category-bar-color)`,
      width: `${(category.value / maximumValue) * 100}%`
    };

    const cssClasses = {
      'as-category-widget__category': true,
      'as-category-widget__category--not-selected': isAnyCategorySelected && (!isSelected || isOther),
      'as-category-widget__category--other': isOther,
      'as-category-widget__category--selected': isSelected
    };

    return (
      <li class={cssClasses} onClick={() => this._toggleCategory(category)}>
        <p class='as-category-widget__info as-body'>
          <div class='as-category-widget__title'>{category.name}</div>
          {readableNumber(category.value)}
        </p>

        <div class='as-category-widget__bar'>
          <div class='as-category-widget__bar-value' style={progressStyles}></div>
        </div>
      </li>
    );
  }

  private _renderOtherCategory(category: Category, options: CategoryOptions) {
    const categoryData = category || {
      name: 'Other',
      value: this._getCategoriesTotalValue(
        this.categories.slice(this.visibleCategories, this.categories.length)
      )
    };

    return this._renderCategory(categoryData, { maximumValue: options.maximumValue, isOther: true });
  }

  private _renderFooter() {
    const selectedCount = this.selectedCategories.length;

    return (
      <footer class='as-category-widget__footer'>
        <div class='as-category-widget__count as-body'>{selectedCount || 'All'} selected</div>
        {this.showClearButton && (
          <button
            class='as-btn as-btn--primary as-btn--s as-category-widget__clear'
            disabled={!selectedCount}
            onClick={() => this.clearSelection()}
          >
            Clear selection
          </button>
        )}
      </footer>
    );
  }

  private _isSelected(categoryName: string) {
    return this.selectedCategories.includes(categoryName);
  }

  private _toggleCategory(category: Category) {
    if (this.disableInteractivity) {
      return;
    }

    this.selectedCategories = this._isSelected(category.name)
      ? this.selectedCategories.filter((currentCategory) => currentCategory !== category.name)
      : [...this.selectedCategories, category.name];

    this._onCategoriesChanged();
  }

  private _onCategoriesChanged() {
    this.categoriesSelected.emit(this.selectedCategories);
  }

  private _getCategoriesMaximumValue(categories: Category[], otherCategoryPresent: boolean = false) {
    return this._getVisibleCategories(categories, otherCategoryPresent).reduce(
      (maximum, currentCategory: Category) => currentCategory.value > maximum ? currentCategory.value : maximum, 0
    );
  }

  private _getCategoriesTotalValue(categories: object[]) {
    return categories.reduce(
      (sum, currentCategory: Category) => currentCategory.value + sum, 0
    );
  }

  private _getBarColor(color: string, options: { isSelected?: boolean, isOther?: boolean } = {}) {
    if (options.isOther) {
      return OTHER_CATEGORY_COLOR;
    }

    if (options.isSelected) {
      return shadeOrBlend(-0.16, color);
    }

    return color;
  }

  private _parseCategories() {
    const otherCategory = this.categories.find(
      (category: Category) => category.name === OTHER_CATEGORY_NAME
    ) as Category;

    if (otherCategory) {
      const categories = this.categories
        .filter((category: Category) => category.name !== otherCategory.name) as Category[];

      return { categories, otherCategory };
    }

    return { categories: this.categories as Category[] };
  }

  private _getVisibleCategories(parsedCategories, otherCategoryPresent: boolean) {
    if (otherCategoryPresent) {
      return parsedCategories;
    }

    return parsedCategories.slice(0, this.visibleCategories);
  }
}

export interface Category {
  name: string;
  value: number;
  color?: string;
}
export interface CategoryOptions {
  maximumValue: number;
  isOther?: boolean;
}
