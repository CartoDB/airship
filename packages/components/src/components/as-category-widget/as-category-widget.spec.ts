import { TestWindow } from '@stencil/core/dist/testing';
import { Category, CategoryWidget } from './as-category-widget';

describe('as-category-widget', () => {
  let categoryWidget;

  beforeEach(() => {
    categoryWidget = new CategoryWidget();
  });

  it('should build', () => {
    expect(new CategoryWidget()).toBeTruthy();
  });

  describe('Rendering', () => {
    let element: HTMLAsCategoryWidgetElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [CategoryWidget],
        html: '<as-category-widget></as-category-widget>'
      });
    });

    it('should render properly', async () => {
      element.heading = 'Category Widget Example';
      element.description = 'Description for Category Widget';
      element.categories = exampleCategories;
      element.showClearButton = true;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });

    it('should not render header when showHeader is false', async () => {
      element.heading = 'Category Widget Example';
      element.description = 'Description for Category Widget';
      element.showHeader = false;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });

    it('should render clear button', async () => {
      element.showClearButton = true;
      element.categories = exampleCategories;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });

    it('should render bars according to total value', async () => {
      element.categories = exampleCategories;
      element.useTotalPercentage = true;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });
  });

  describe('Behaviour', () => {
    let element: HTMLAsCategoryWidgetElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [CategoryWidget],
        html: '<as-category-widget></as-category-widget>'
      });
    });

    it('should emit an event containing selected categories when a new category is selected', async () => {
      element.categories = exampleCategories;

      const categoriesSelectedSpy = jest.fn();
      element.addEventListener('categoriesSelected', categoriesSelectedSpy);
      await testWindow.flush();

      const categoryElement = element.querySelector('.as-category-widget__category') as HTMLDivElement;
      categoryElement.click();

      expect(categoriesSelectedSpy).toHaveBeenCalled();
      expect(categoriesSelectedSpy.mock.calls[0][0].detail).toEqual([exampleCategories[0].name]);
    });

    it('should emit an event containing selected categories when categories are cleared', async () => {
      element.categories = exampleCategories;
      element.showClearButton = true;
      await testWindow.flush();

      const categoryElement = element.querySelector('.as-category-widget__category') as HTMLDivElement;
      categoryElement.click();

      await testWindow.flush();

      const spy = jest.fn();
      element.addEventListener('categoriesSelected', spy);

      const clearButton = element.querySelector('.as-category-widget__clear') as HTMLButtonElement;
      clearButton.click();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls[0][0].detail).toEqual([]);
    });
  });

  describe('.getSelectedCategories', () => {
    it('should return selectedCategories', () => {
      const selectedCategories = ['Category 1'];
      categoryWidget.selectedCategories = selectedCategories;

      expect(categoryWidget.getSelectedCategories()).toEqual(selectedCategories);
    });
  });

  describe('.clearSelection', () => {
    beforeEach(() => {
      spyOn(categoryWidget, '_onCategoriesChanged');
    });

    it('should clear selected categories', () => {
      const selectedCategories = ['Category 1'];
      categoryWidget.selectedCategories = selectedCategories;

      categoryWidget.clearSelection();

      expect(categoryWidget.getSelectedCategories()).toEqual([]);
    });
  });

  describe('._isSelected', () => {
    it('should return true if category is selected', () => {
      const category = 'Category 1';
      categoryWidget.selectedCategories = [category];

      expect(categoryWidget._isSelected(category)).toBe(true);
    });

    it('should return false if category is not selected', () => {
      expect(categoryWidget._isSelected('Fake Category')).toBe(false);
    });
  });

  describe('._toggleCategory', () => {
    beforeEach(() => {
      spyOn(categoryWidget, '_onCategoriesChanged');
    });

    it('should select category if it is not selected', () => {
      const category = { name: 'Category 1', value: 10, color: '#000' };
      categoryWidget._toggleCategory(category);

      expect(categoryWidget.getSelectedCategories()).toEqual([category.name]);
    });

    it('should remove category if it was selected', () => {
      const category = { name: 'Category 1', value: 10, color: '#000' };
      categoryWidget.selectedCategories = [category.name];

      categoryWidget._toggleCategory(category);

      expect(categoryWidget.getSelectedCategories()).toEqual([]);
    });

    it('should not toggle category if interaction is disabled', () => {
      categoryWidget.disableInteractivity = true;

      const category = { name: 'Category 1', value: 10, color: '#000' };
      categoryWidget._toggleCategory(category);

      expect(categoryWidget.getSelectedCategories()).toEqual([]);
    });
  });

  describe('._getCategoriesMaximumValue', () => {
    it('should return maximum value of visible categories', () => {
      categoryWidget.useTotalPercentage = false;
      categoryWidget.visibleCategories = 5;

      expect(categoryWidget._getCategoriesMaximumValue(exampleCategories)).toEqual(1000);
    });
  });

  describe('._getCategoriesTotalValue', () => {
    it('should return value sum of passed categories', () => {
      expect(categoryWidget._getCategoriesTotalValue(exampleCategories)).toEqual(6100);
    });
  });

  describe('._getBarColor', () => {
    it('should return provided color if category is not selected', () => {
      expect(categoryWidget._getBarColor('#FFFFFF')).toBe('#FFFFFF');
    });

    it('should return darker color if category is selected', () => {
      expect(categoryWidget._getBarColor('#FFFFFF', { isSelected: true })).toBe('#d6d6d6');
    });
  });
});

const exampleCategories: Category[] = [
  { name: 'Bars & Restaurants', value: 1000, color: '#FABADA' },
  { name: 'Fashion', value: 900 },
  { name: 'Grocery', value: 800 },
  { name: 'Health', value: 400 },
  { name: 'Shopping mall', value: 250 },
  { name: 'Transportation', value: 1550 },
  { name: 'Leisure', value: 1200 }
];
