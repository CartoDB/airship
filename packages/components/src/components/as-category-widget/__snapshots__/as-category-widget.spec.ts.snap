// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-category-widget Rendering should format display value when formatValue prop is present 1`] = `
<as-category-widget
  class="hydrated"
>
  
  <h2
    class="as-category-widget__heading"
  >
    
  </h2>
  <p
    class="as-category-widget__description as-body"
  >
    
  </p>
  <ul
    class="as-category-widget__list"
  >
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Bars & Restaurants
        </div>
        1000€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="background-color: rgb(250, 186, 218); width: 64.51612903225806%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Fashion
        </div>
        900€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 58.06451612903226%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Grocery
        </div>
        800€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 51.61290322580645%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Health
        </div>
        400€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 25.806451612903224%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Shopping mall
        </div>
        250€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 16.129032258064516%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Transportation
        </div>
        1550€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 100%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Leisure
        </div>
        1200€
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 77.41935483870968%;"
        />
      </div>
    </li>
    
  </ul>
  <footer
    class="as-category-widget__footer"
  >
    <div
      class="as-category-widget__count as-body"
    >
      All selected
    </div>
    
  </footer>
</as-category-widget>
`;

exports[`as-category-widget Rendering should not render header when showHeader is false 1`] = `
<as-category-widget
  class="hydrated"
>
  
  
  <ul
    class="as-category-widget__list"
  >
    
  </ul>
  <footer
    class="as-category-widget__footer"
  >
    <div
      class="as-category-widget__count as-body"
    >
      All selected
    </div>
    
  </footer>
</as-category-widget>
`;

exports[`as-category-widget Rendering should render bars according to total value 1`] = `
<as-category-widget
  class="hydrated"
>
  
  <h2
    class="as-category-widget__heading"
  >
    
  </h2>
  <p
    class="as-category-widget__description as-body"
  >
    
  </p>
  <ul
    class="as-category-widget__list"
  >
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Bars & Restaurants
        </div>
        1.0K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="background-color: rgb(250, 186, 218); width: 16.39344262295082%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Fashion
        </div>
        900
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 14.754098360655737%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Grocery
        </div>
        800
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 13.114754098360656%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Health
        </div>
        400
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 6.557377049180328%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Shopping mall
        </div>
        250
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 4.098360655737705%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Transportation
        </div>
        1.6K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 25.40983606557377%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Leisure
        </div>
        1.2K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 19.672131147540984%;"
        />
      </div>
    </li>
    
  </ul>
  <footer
    class="as-category-widget__footer"
  >
    <div
      class="as-category-widget__count as-body"
    >
      All selected
    </div>
    
  </footer>
</as-category-widget>
`;

exports[`as-category-widget Rendering should render clear button 1`] = `
<as-category-widget
  class="hydrated"
>
  
  <h2
    class="as-category-widget__heading"
  >
    
  </h2>
  <p
    class="as-category-widget__description as-body"
  >
    
  </p>
  <ul
    class="as-category-widget__list"
  >
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Bars & Restaurants
        </div>
        1.0K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="background-color: rgb(250, 186, 218); width: 64.51612903225806%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Fashion
        </div>
        900
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 58.06451612903226%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Grocery
        </div>
        800
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 51.61290322580645%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Health
        </div>
        400
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 25.806451612903224%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Shopping mall
        </div>
        250
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 16.129032258064516%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Transportation
        </div>
        1.6K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 100%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Leisure
        </div>
        1.2K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 77.41935483870968%;"
        />
      </div>
    </li>
    
  </ul>
  <footer
    class="as-category-widget__footer"
  >
    <div
      class="as-category-widget__count as-body"
    >
      All selected
    </div>
    <button
      class="as-btn as-btn--primary as-btn--s as-category-widget__clear"
      disabled=""
    >
      Clear selection
    </button>
  </footer>
</as-category-widget>
`;

exports[`as-category-widget Rendering should render properly 1`] = `
<as-category-widget
  class="hydrated"
>
  
  <h2
    class="as-category-widget__heading"
  >
    Category Widget Example
  </h2>
  <p
    class="as-category-widget__description as-body"
  >
    Description for Category Widget
  </p>
  <ul
    class="as-category-widget__list"
  >
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Bars & Restaurants
        </div>
        1.0K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="background-color: rgb(250, 186, 218); width: 64.51612903225806%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Fashion
        </div>
        900
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 58.06451612903226%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Grocery
        </div>
        800
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 51.61290322580645%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Health
        </div>
        400
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 25.806451612903224%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Shopping mall
        </div>
        250
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 16.129032258064516%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Transportation
        </div>
        1.6K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 100%;"
        />
      </div>
    </li>
    <li
      class="as-category-widget__category"
    >
      <p
        class="as-category-widget__info as-body"
      >
        <div
          class="as-category-widget__title"
        >
          Leisure
        </div>
        1.2K
      </p>
      <div
        class="as-category-widget__bar"
      >
        <div
          class="as-category-widget__bar-value"
          style="width: 77.41935483870968%;"
        />
      </div>
    </li>
    
  </ul>
  <footer
    class="as-category-widget__footer"
  >
    <div
      class="as-category-widget__count as-body"
    >
      All selected
    </div>
    <button
      class="as-btn as-btn--primary as-btn--s as-category-widget__clear"
      disabled=""
    >
      Clear selection
    </button>
  </footer>
</as-category-widget>
`;
